### PR TITLE
[Security/Extension] Extension Authentication Backend

### PR DESCRIPTION
**Pull request by RyanL1997**
### Description
Extension Authentication Backend

Some additional modification in JWTVendor:
- Add the support of BWCMode
~~- _BWCMode On_: plain text roles will be passed into claim. It also accepts nonString configuration, but it will be converted into `string` as the payload. The roles information will be stored in the payload field `dec_r` (standing for decrypted roles).~~
~~- _BMCMode Off_: encrypted roles will be passed into claim, and it will always be a encrypted `string`. The roles information will be stored in the payload field `enc_r` (standing for encrypted roles).~~

### Category
New feature

### Issues Resolved
* Resolve: #2619

### Check List
- [x] New functionality includes testing
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by RyanL1997** [comment]
Testing is WIP

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (scrawfor99) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:84]: Should this throw an error instead of just logging?

File: src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java
Comment (RyanL1997) [reply src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java:831]: AC is standing for AuthCredentials. Nice to point it out. I will also including that in the terminology list I'm creating.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:260]: In the original `HTTPJwtAuthenticator.java` it does the same thing. I believe the reason to do this is that we also need to handle the case for some nonString claims. In the test case, we also have a test case called `testNonStringClaim`. I think this scenario only happens when the user enable the BWCMode, which means instead of passing encoded roles, plain text roles will be passed into the claim. For the disabling of BMCMode, the role will always be encrypted, so that it will always be an encrypted `string`.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:81]: Yes this will directly skip the `else` and throw another error log L110 with an runtime exception at [L112](https://github.com/opensearch-project/security/pull/2672/files#diff-efd217975f23af4bb84808e2771bfc85211f12b59bb619da430a3363902207ceR112). The way we want to do it in this way is because we would like to log some specific information for user to know the reason of failure creation of this token. There are also some `try`/`catch` statement in the `else` for checking encryption algorithm, and each of them will work in the same pattern. It will tell the user why they failed first by these error log and give them the runtime exception error at L112.

**Comment by RyanL1997** [comment]
I'm working on the fix of just 1 test case: `testRolesArray` and some code hygiene issue. Other than that this PR is ready for review.

**Comment by RyanL1997** [comment]
The CI will be fixed after the merge of https://github.com/opensearch-project/security/pull/2715

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (cwperks) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:151]: Does this mean that this takes anything passed in the `Authorization` header as long as it doesn't contain the word `Basic` anywhere? In the next block if it contains the word `bearer` it will strip that out and trim the rest?

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (cwperks) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:196]: The null check isn't needed here or in the else if, this will never be null.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (cwperks) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:260]: The message here shouldn't contain the word `roles`. Can you update the message to `Expected type String for subject...`

This should be updated in the `HTTPJwtAuthenticator` as well. There is overlap between here and the `HTTPJwtAuthenticator` class. Maybe we can create a utility class with static methods or maybe add these to an abstract class that both inherit from?

File: src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java
Comment (cwperks) [inline src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java:831]: Should this be behind the extensions feature flag?

**Review by cwperks — COMMENTED** [review]
Hey @RyanL1997 , I took a first pass at this and like the progress and tests that have been added here. Do you have any thoughts about configuration of this backend? Is this a special backend that if extensions is enabled then is evaluated before the list of auth domains in the `authc` section of the security plugin `config.yml` file? Have you been able to install the security plugin with this change into an opensearch node and show the change in use?

**Review by scrawfor99 — COMMENTED** [review]
Hi @RyanL1997, things look very good here. I would be happy to approve if you are able to address Craig's comments. I think he covered everything.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:184]: Yes, I think this is an good idea. Updating it now.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:63]: Yes, I will update this~

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Done~

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Good catch. That was a redundancy. I have removed this.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:155]: Yes, I think it is also work with the `.match()`. I was following the same expression in `HTTPJwtAuthenticator.java` ([L170](https://github.com/opensearch-project/security/blob/6997f97f950683c1125ac5b0bb797c40fe8877e8/src/main/java/com/amazon/dlic/auth/http/jwt/HTTPJwtAuthenticator.java#L170)). So, maybe we should just leave like this.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:12]: What are your thoughts on this package, com.amazon doesn't make much sense to me.  I'd suggest a move this under a `org.opensearch.security.auth` package or child package.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:53]: Why does this authenticating backend check for BASIC auth headers, shouldn't it only support bearer auth?

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:73]: If any extensions needs the BWC mode all extensions will get roles information? What do you think about making this determination when the token is generated instead?

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:129]: Is their a specific action that needs the permission?  The InternalAuthenticationBackend doesn't seem to need this PrivilegedAction, what is this class doing?

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:136]: Take a look at these classes from @scrawfor99 's PR, they do a good job isolating the specific functionality from the overall context.  While you might not need extra classes here, I think the parsing extracting logic would be cleaner

- [RestTokenExtractor](https://github.com/opensearch-project/OpenSearch/blob/e2269cbd40eeb44a9062991090387ca543639fef/server/src/main/java/org/opensearch/identity/tokens/RestTokenExtractor.java)
- [BasicAuthToken](https://github.com/opensearch-project/OpenSearch/blob/e2269cbd40eeb44a9062991090387ca543639fef/server/src/main/java/org/opensearch/identity/tokens/BasicAuthToken.java)

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:229]: Instead lets return false like in the different proxy authenticators

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:49]: While extensions will use this functionality.  This class is a delegate authenticating backend, lets remove the name extension and reframe to its purpose, what do you think of this?

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (DarshitChanpura) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:202]: is `audience` not mandatory here? If not, why?

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (DarshitChanpura) [inline src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java:234]: should this hard-coded value be part of Constants file?

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Fixed. That makes sense to me too. I was doing this because I was following the class of `HTTPJwtAuthenticator`.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: That was the question I raised up during one of our stand-up meetings. I assume that the `HTTPJwtAuthenticator` were doing this could be due to leniency in handling to support non-standard header formats that do not include the “Bearer” keyword explicitly. However, for our case, we are require the keyword "Bearer" for our auth header. I will fix this in the next commit.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Correct, that `getPublicKey()` cannot handle the case for Base64 encoding.

File: src/main/java/org/opensearch/security/authtoken/jwt/JwtVendor.java
Comment (scrawfor99) [inline src/main/java/org/opensearch/security/authtoken/jwt/JwtVendor.java:172]: I am not sure what "er" means. I know there are some conflicting opinions, but I am of the mind that we lose meaning at a certain point of abbreviating things. I am guessing this stands for "____ roles"? Perhaps we can make this a more meaningful key?

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Thanks for the suggestion! Already fixed

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Thanks for the suggestions, fixed~

**Review by scrawfor99 — APPROVED** [review]
Looks good Ryan! Great job :)

File: src/main/java/org/opensearch/security/authtoken/jwt/JwtVendor.java
Comment (DarshitChanpura) [reply src/main/java/org/opensearch/security/authtoken/jwt/JwtVendor.java:172]: i think `er` stands for `encrypted roles`.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: I think we can leave like this. I was following the pattern of [HTTPJwtAuthenticator](https://github.com/opensearch-project/security/blob/main/src/main/java/com/amazon/dlic/auth/http/jwt/HTTPJwtAuthenticator.java#L219) and [HTTPBasicAuthenticator](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/http/HTTPBasicAuthenticator.java#L77).

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Fixed!

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: @peternied and @DarshitChanpura , Thanks for the information and suggestions. Yes I think this is a good idea. We can do that as the follow-up PR for linking this backend with the configuration setup.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Good suggestion! Fixed.

File: src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/com/amazon/dlic/auth/http/jwt/HTTPExtensionJwtAuthenticator.java]: Good catch! I switched that to a `else`.

File: src/main/java/org/opensearch/security/http/HTTPOnBehalfOfJwtAuthenticator.java
Comment (cwperks) [inline src/main/java/org/opensearch/security/http/HTTPOnBehalfOfJwtAuthenticator.java:193]: nit: Can we change this variable name to `decryptedRoles`?

**Review by cwperks — COMMENTED** [review]
Thank you @RyanL1997! I left 2 comments that are pretty minor. Once addressed I will approve this PR and we can merge it into the feature branch.

File: src/main/java/org/opensearch/security/http/HTTPOnBehalfOfJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/org/opensearch/security/http/HTTPOnBehalfOfJwtAuthenticator.java]: Sure, I have fixed.

File: src/main/java/org/opensearch/security/http/HTTPOnBehalfOfJwtAuthenticator.java
Comment (RyanL1997) [reply src/main/java/org/opensearch/security/http/HTTPOnBehalfOfJwtAuthenticator.java]: Fixed

**Review by cwperks — APPROVED** [review]

**Review by DarshitChanpura — APPROVED** [review]
Nicely done!

**Comment by RyanL1997** [comment]
Merging, and the CI is failing due to the new introduced term check for 'extension' at previous PR: Extensions config for JWT signing/encryption key (#2671). I have created a follow-up issue at: https://github.com/opensearch-project/security/issues/2760

**Pull request merged**
